src/hotspot/share/memory/virtualspace.cpp

```diff
@@ -239,34 +240,31 @@ void ReservedSpace::reserve(size_t size,
     }
     // Always return, not possible to fall back to reservation not using a file.
     return;
-  } else if (use_explicit_large_pages(page_size)) {
+  }
+
+  // == Case 2 ==
+  if (use_explicit_large_pages(page_size)) {
     // System can't commit large pages i.e. use transparent huge pages and
     // the caller requested large pages. To satisfy this request we use
     // explicit large pages and these have to be committed up front to ensure
     // no reservations are lost.
-    size_t used_page_size = page_size;
-    char* base = NULL;
-
     do {
-      base = reserve_memory_special(requested_address, size, alignment, used_page_size, executable);
+      char* base = reserve_memory_special(requested_address, size, alignment, page_size, executable);
       if (base != NULL) {
-        break;
+        // Successful reservation using large pages.
+        initialize_members(base, size, alignment, page_size, true, executable);
+        return;
       }
-      used_page_size = os::page_sizes().next_smaller(used_page_size);
-    } while (used_page_size > (size_t) os::vm_page_size());
+      page_size = os::page_sizes().next_smaller(page_size);
+    } while (page_size > (size_t) os::vm_page_size());
 
-    if (base != NULL) {
-      // Successful reservation using large pages.
-      initialize_members(base, size, alignment, used_page_size, true, executable);
-      return;
-    }
     // Failed to reserve explicit large pages, do proper logging.
     log_on_large_pages_failure(requested_address, size);
     // Now fall back to normal reservation.
-    page_size = os::vm_page_size();
```

Comment:
In the original code the setting of `page_size` appears to relate to the comment about falling back, but it would appear that comment is really just about falling through to the case three code - is that right? If so the assertion would be a little clearer if put directly after the do-while loop (though it's really validating the loop termination condition anyway).

Reply:
> In the original code the setting of page_size appears to relate to the comment about falling back, but it would appear that comment is really just about falling through to the case three code - is that right?

I'd say it's both -- we are falling to Case 3 _and_ `page_size` needs to have a suitable value for Case 3.

Reply:
Ok - no big deal.